Jellyfin.Server.Implementations/MediaSegments/MediaSegmentsManager.cs

```diff
@@ -0,0 +1,213 @@
+    /// <param name="found">The found segment.</param>
+    /// </summary>
+    private void AddOrUpdateSegment(JellyfinDbContext dbContext, MediaSegment segment, MediaSegment? found)
+    {
+        if (found != null)
+        {
+            found.StartTicks = segment.StartTicks;
+            found.EndTicks = segment.EndTicks;
+            found.Action = segment.Action;
+            found.Comment = segment.Comment;
+        }
+        else
+        {
+            dbContext.Segments.Add(segment);
+        }
+    }
+
+    /// <summary>
+    /// Validate a segment: itemId, start >= end and type.
+    /// </summary>
+    /// <param name="segment">The segment to validate.</param>
+    private void ValidateSegment(MediaSegment segment)
+    {
+        if (segment.ItemId.Equals(default))
+        {
+            throw new ArgumentException($"itemId is default: itemId={segment.ItemId} for segment with type '{segment.Type}.{segment.TypeIndex}'");
+        }
+
+        if (segment.StartTicks >= segment.EndTicks)
+        {
+            throw new ArgumentException($"start >= end: {segment.StartTicks}>={segment.EndTicks} for segment itemId '{segment.ItemId}' with type '{segment.Type}.{segment.TypeIndex}'");
+        }
+    }
+
+    /// <summary>
+    /// Delete all segments when itemid is deleted from library.
+    /// </summary>
+    /// <param name="sender">The sending entity.</param>
+    /// <param name="itemChangeEventArgs">The <see cref="ItemChangeEventArgs"/>.</param>
+    private async void LibraryManagerItemRemoved(object? sender, ItemChangeEventArgs itemChangeEventArgs)
+    {
+        await DeleteSegments(itemChangeEventArgs.Item.Id).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async Task<IReadOnlyList<MediaSegment>> DeleteSegments(Guid itemId, int? streamIndex = null, int? typeIndex = null, MediaSegmentType? type = null)
+    {
+        List<MediaSegment> allSegments;
+
+        if (itemId.IsEmpty())
+        {
+            throw new ArgumentException("Default value provided", nameof(itemId));
+        }
+
+        var dbContext = await _dbProvider.CreateDbContextAsync().ConfigureAwait(false);
+        await using (dbContext.ConfigureAwait(false))
+        {
+            IQueryable<MediaSegment> queryable = dbContext.Segments.Where(s => s.ItemId.Equals(itemId));
+
+            if (streamIndex is not null)
+            {
+                queryable = queryable.Where(s => s.StreamIndex == streamIndex);
+            }
+
+            if (type is not null)
+            {
+                queryable = queryable.Where(s => s.Type == type);
+            }
+
+            if (typeIndex is not null)
+            {
+                queryable = queryable.Where(s => s.TypeIndex == typeIndex);
+            }
+
+            allSegments = await queryable.ToListAsync().ConfigureAwait(false);
+
+            dbContext.Segments.RemoveRange(allSegments);
+            await dbContext.SaveChangesAsync().ConfigureAwait(false);
```

Comment:
Is the return value of this method used? If not, you can save a DB round trip by calling `ExecuteDeleteAsync`